w11h03/test/pgdp/networking/SocketTest.java

```diff
@@ -0,0 +1,455 @@
+        String message = """
+                    I tell you what, this has been really fun.
+                    And I got to help make something pretty cool, so I’ve got no complaints.
+                    I mean, not me, exactly, but close enough.
+                    It’s the kind of thing that makes you glad you stopped and smelled the pine trees along the way, you know?""";
+        byte[] messageBuffer = StandardCharsets.UTF_8.encode(message).array();
+        Thread connectThread = new Thread(this::connect);
+        connectThread.start();
+        Socket client = server.accept();
+        client.getOutputStream().write(SERVER_HELLO);
+        client.getOutputStream().flush();
+        getAllBytes(connectThread, client);
+
+        Thread clientThread = new Thread(() -> {
+            dataHandler.sendMessage(message);
+        });
+        // Send Server Hello
+        clientThread.start();
+        Thread.sleep(SOCKET_TIMEOUT);
+        // Write ACK
+        var buffer = getAllBytes(clientThread, client);
+        assertEquals(0x01, buffer[0], "First byte");
+        assertEquals(messageBuffer.length, bufferToInt(copyOf(buffer, 1, 3, "Message length")), "Message length");
+
+        var expected = new byte[3 + messageBuffer.length];
+        expected[0] = 0x01;
+        expected[1] = (byte) 1;
+        expected[2] = (byte) 35;
+        System.arraycopy(messageBuffer, 0, expected, 3, expected.length - 3);
+        assertArrayEquals(expected, buffer);
+
+        // Assert end of output
+        assertEquals(expected.length, buffer.length, "Transferred bytes did not end after expected message.");
+    }
+
+    @Test
+    @DisplayName("sendMessage() – huge")
+    void testMessagesHuge() throws IOException, InterruptedException {
+        // Ideally, we would use resources here, but that seems difficult with the current tests setup
+        String lipsum = Files.readString(Path.of("./test/pgdp/networking/lipsum.txt"));
+        String lipsumTruncated = Files.readString(Path.of("./test/pgdp/networking/lipsum_truncated.txt"));
+        Thread connectThread = new Thread(this::connect);
+        connectThread.start();
+        Socket client = server.accept();
+        client.getOutputStream().write(SERVER_HELLO);
+        client.getOutputStream().flush();
+        getAllBytes(connectThread, client);
+
+        Thread clientThread = new Thread(() -> {
+            dataHandler.sendMessage(lipsum);
+        });
+        // Send Server Hello
+        clientThread.start();
+        Thread.sleep(SOCKET_TIMEOUT);
+        // Write ACK
+        var buffer = getAllBytes(clientThread, client);
+        assertEquals(0x01, buffer[0], "First byte");
+        int messageLength = bufferToInt(copyOf(buffer, 1, 3, "Message length"));
+        assertEquals(0xffff, messageLength, "Message length was incorrect.");
+        assertEquals(lipsumTruncated, new String(buffer).substring(3), "Text message was incorrect.");
```

Comment:
This test fails on platforms where the default charset is not UTF-8

```suggestion
        assertEquals(lipsumTruncated, new String(buffer, StandardCharsets.UTF_8).substring(3), "Text message was incorrect.");
```